### PR TITLE
Update GSgnnLinkPredictionModelInterface and support mini-batch based full graph inference

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/gnn.py [hunk 1]
```diff
@@ -673,7 +678,14 @@ def get_input_embeds(input_nodes):
                 input_nodes = {data.g.ntypes[0]: input_nodes}
             return {ntype: input_embeds[ntype][ids].to(device) \
                     for ntype, ids in input_nodes.items()}
-        embeddings = model.gnn_encoder.dist_inference(data.g, get_input_embeds,
+        if minibatch:
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:681]: we should have a function called `mini_batch_inference` outside `full_graph_inference`

File: python/graphstorm/model/gnn_encoder_base.py [hunk 2]
```diff
@@ -104,6 +104,83 @@ def dist_inference(self, g, get_input_embeds, batch_size, fanout,
         return dist_inference(g, self, get_input_embeds, batch_size, fanout,
                               edge_mask=edge_mask, task_tracker=task_tracker)
 
+
+def dist_minibatch_inference(g, gnn_encoder, get_input_embeds, batch_size, fanout,
```
Comment (zheng-da) [inline python/graphstorm/model/gnn_encoder_base.py:108]: why do we want to have mini-batch inference on the entire graph?

File: python/graphstorm/model/gnn.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/model/gnn.py:681]: It is still full graph inference, but in a mini-batch manner.

File: python/graphstorm/model/gnn_encoder_base.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/model/gnn_encoder_base.py:108]: If there is a custom model, we may not be able to do layer-by-layer full graph inference.
That's why we need mini-batch inference.

File: python/graphstorm/inference/ep_infer.py [hunk 3]
```diff
@@ -77,7 +77,8 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
         sys_tracker.check('start inferencing')
         self._model.eval()
         embs = do_full_graph_inference(self._model, loader.data, fanout=loader.fanout,
-                                       task_tracker=self.task_tracker)
+                                       task_tracker=self.task_tracker,
+                                       minibatch=use_mini_batch_infer)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/inference/ep_infer.py:81]: Now we have two ways to do full graph inference: 1) layer-by-layer inference and 2) node-batch by node-batch (mini-batch here) inference.

[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/inference/ep_infer.py:81]: my question is why we do mini-batch inference in the function called `do_full_graph_inference`.

File: python/graphstorm/model/gnn.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/model/gnn.py:681]: is there any advantage of computing GNN embeddings of all nodes in a graph?

File: python/graphstorm/model/gnn_encoder_base.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/model/gnn_encoder_base.py:108]: i see.

File: python/graphstorm/inference/lp_infer.py [hunk 4]
```diff
@@ -43,6 +43,7 @@ class GSgnnLinkPredictionInfer(GSInfer):
     # TODO(zhengda) We only support full-graph inference for now.
     def infer(self, data, loader, save_embed_path,
             edge_mask_for_gnn_embeddings='train_mask',
+            use_mini_batch_infer=False,
```
Comment (zheng-da) [inline python/graphstorm/inference/lp_infer.py:46]: should we have this as an argument? if mini-batch inference is used for custom models, we can easily test if the input model is a custom model.

File: python/graphstorm/model/gnn.py [hunk 5]
```diff
@@ -622,7 +623,7 @@ def loss_func(self):
         return self._loss_fn
 
 def do_full_graph_inference(model, data, batch_size=1024, fanout=None, edge_mask=None,
-                            task_tracker=None):
+                            task_tracker=None, minibatch=False):
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:626]: i think it's better to not invoke mini-batch inference code inside full-graph inference.

**Review by zheng-da — APPROVED** [review]
overall, it looks good to me. We need to be careful of using mini-batch inference when we need to compute BERT embeddings on nodes.
